app/forms/changeGroupLead/NewGroupLeadEnterContactAddressFormProvider.scala

```diff
@@ -16,13 +16,71 @@
 
 package forms.changeGroupLead
 
+import forms.changeGroupLead.NewGroupLeadEnterContactAddressFormProvider._
+import forms.mappings.Mappings
+import models.changeGroupLead.NewGroupLeadAddressDetails
 import play.api.data.Form
-import play.api.data.Forms.text
+import play.api.data.Forms._
+import play.api.data.validation.Constraint
+import play.api.data.validation.Constraints.pattern
 
-class NewGroupLeadEnterContactAddressFormProvider {
-  def apply(): Form[String] =
-    Form("value" -> text)
-}
+import javax.inject.Inject
+
+//todo: add more validation: see SoT
+class NewGroupLeadEnterContactAddressFormProvider @Inject() extends Mappings {
+
+   def apply(): Form[NewGroupLeadAddressDetails] = Form(
+     mapping(
+       addressLine1 -> text("newGroupLeadEnterContactAddress.error.addressLine.required")
+        .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine1.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine1.invalid.line")),
+       addressLine2 -> text("newGroupLeadEnterContactAddress.error.addressLine.required")
+        .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine2.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine2.invalid.line")),
+       addressLine3 -> optional(
+         text("newGroupLeadEnterContactAddress.error.AddressLine2.required")
+           .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine3.length"))
+           .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine3.invalid.line"))
+       ),
+       addressLine4 -> text("newGroupLeadEnterContactAddress.error.addressLine4.required")
+         .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.addressLine4.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.addressLine4.invalid.line")),
+       postalCode -> optional(
+         text("newGroupLeadEnterContactAddress.error.postalCode.required ")
+           .verifying(maxLength(8, "newGroupLeadEnterContactAddress.error.postalCode.inRange"))
+           .verifying(minLength(5, "newGroupLeadEnterContactAddress.error.postalCode.inRange"))
+           .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.postalCode.inRange"))
+       ),
+       countryCode -> text("newGroupLeadEnterContactAddress.error.countryCode.required")
+         .verifying(maxLength(maxAddressLineLength, "newGroupLeadEnterContactAddress.error.countryCode.length"))
+         .verifying(isValidAddressLine("newGroupLeadEnterContactAddress.error.countryCode.invalid"))
+    )(NewGroupLeadAddressDetails.apply)(NewGroupLeadAddressDetails.unapply).verifying(
+       "newGroupLeadEnterContactAddress.error.postalCode.required",
```

Comment:
this may need to be done with Constraint so that you can pass through an arg of countryCode so it links back

Reply:
yes, this was just an experiment as I do not know how to pass the field name here. Probably I will follow reg example